### ISSUE TITLE
Add DNS server to VM and minishift commands to mange its life cycle

### ISSUE THREAD

**Issue by hferentschik** [minishift/minishift#201]
We could extend minishift with some commands to mange a very simplistic DNS server which returns the IP of the VM for all wildcard queries matching the configured OpenShift default domain. For all other cases, it can forward to an upstream DNS server. The DNS process could run on the host or maybe even easier in the VM. The commands I am thinking of are 'dns-start', 'dns-stop', 'dns-status'. Some commands could be made implicit as well, for example 'minishift start' could automatically start the DNS server, in case a non [xip|nip].io IP is used.

The aim is a similar experience which can be achieved via the Vagrant Landrush plugin, except that your solution can be much simpler, since it only target one specific use case. 

The required "tricks" to re-configure the host system to make use of the minishift managed DNS can be learned from Landrush.

**Comment by mohammedzee1000** [comment]
I played around around with this a bit  and i got it working with dnsmasq.

TL;DR : Setup dnsmasq on a centos vm (repersenting the minishift node) and add dns entry to resolve wildcard dns (say .example.com=<ip of minishift vm>) and then configure the base machine to use the minishift vm as one of its DNS server.

Steps i followed : 

**On the centos vm :** 

`$ yum -y install dnsmasq`

```
$ cat /etc/dnsmasq.conf

...
listen-address=192.168.122.10
...
server 8.8.8.8

```

```
$ cat /etc/dnsmasq.d/dns.hosts

domain=example.com, 192.168.122.0/24, local
address=/.example.com/192.168.122.10
bind-interfaces
```

**On linux client System:** 

`$ nmcli con mod mycon ipv4.dns "192.168.122.10"`

**On windows client system:**

`$ netsh interface ip add dns “mycon” 192.168.122.10 index=1`

The dns is able to resolve after this on both clients (tested with nslookup and ping).

**Comment by mohammedzee1000** [comment]
@gbraad wdyt?

**Comment by gbraad** [comment]
Have been able to replicate this. Using:

```
address=/.192.168.122.10.xip.io/192.168.122.10
address=/.192.168.122.10.nip.io/192.168.122.10
address=/.192.168.122.10.local/192.168.122.10
```

we can catch the resolves to xip/nip and force the expected result. The only problem we might have is with the `0.0.0.0` route when a VPN is used, as the DNS is resolved on a non-localhost address. However, we might consider forwarding or VPNkit for this.

Will set up a small PoC in code to generate the config. Note: using the changes of #1792 we can also force a certain nameserver inside the VM to be used...

**Comment by gbraad** [comment]
when using the VM approach, we can only offer this option for a single profile. At the moment that would not be an issue, as we will release this first as an experimental option, after which we can decide how to offer this on the desktop side. at the moment, we are still a single-binary solution (with the exception of the drivers).

**Comment by budhrg** [comment]
I will co-assist with the issue.

**Comment by gbraad** [comment]
Currently using:

```
$ docker run \
    --name dnsmasq \
    -v $(pwd)/dnsmasq.hosts:/dnsmasq.hosts \
    -p '0.0.0.0:53:5353/udp' \
    -d gbraad/dnsmasq
```

to start a DNS server from the VM. This survives restarts and can be set to use a restart policy if needed. (Note: do not want to modify the ISO unnecessary... especially as the current implementation could only support one instance when hosted from the VM).

**Comment by gbraad** [comment]
Juts throwing it out here; what would be the experience of setting up the VM for offline use?
 /cc: @minishift/minishift-dev

**Comment by gbraad** [comment]
Just throwing it out here; what would be the experience of setting up the VM for offline use?
/cc: @minishift/minishift-dev

**Comment by gbraad** [comment]
@mohammedzee1000 I think the `index=1` is wrong. In testing I see a timeout occurs which lasts 2 seconds, as it tries to use another resolver first (the index=0), so adding as:

```
C:\> netsh interface ip add dns "vEthernet (DockerNAT)" 10.0.75.128 index=0
```

solved this for me. So `index=0` is the actual first nameserver, right?

**Comment by anjannath** [comment]
Following the instructions on https://github.com/minishift/minishift/issues/201#issuecomment-368447443 I was able to replicate this..

Also, the vSwitch `Default Switch` works, and i didn't have to set a static ip, as the `Default Switch` is able to assign ip addresses without relying on a external DHCP server.

**Comment by gbraad** [comment]
On Linux I do not see anything happen to the DNS config:

```
$ sudo nmcli con mod "Wired connection 1" ipv4.dns "192.168.42.172" 
$ nmcli
...
DNS configuration:
        servers: 8.8.8.8 8.8.4.4
        interface: eth0

        servers: 10.0.21.1
        interface: eth2

        servers: 172.26.219.97
        domains: mshome.net
        interface: eth1
...
$ cat /etc/resolv,conf
# Generated by NetworkManager
search mshome.net
nameserver 8.8.8.8
nameserver 8.8.4.4
nameserver 10.0.21.1
# NOTE: the libc resolver may not support more than 3 nameservers.
# The nameservers listed below may not be recognized.
nameserver 172.26.219.97
```

Any ideas?

**Comment by gbraad** [comment]
on macOS we can retrieve the current DNS with:

```
$ networksetup -getdnsservers Wi-Fi
```

and update with:

```
$ networksetup -setdnsservers Wi-Fi 8.8.8.8 [...]
```

**Comment by anjannath** [comment]
@gbraad I add the name server to `/etc/resolv.conf`, then make the file immutable with `chattr +i /etc/resolv.conf` so that DHCP cannot overwrite the file.

**Comment by gbraad** [comment]
I know... but that is a bad idea!?!

**Comment by gbraad** [comment]
solved the issue on macOS. Also, be aware `nslookup` does NOT use `/etc/resolver`... it only deals with `/etc/resolv.conf`

Create individual files under `/etc/resolver`, named like `minishift-192.168.64.13.xip.io` and in the file specify:
```
domain 192.168.64.13.xip.io
```
Note: the filename already forces the domain, but for cleanup purpose I'd rather identify them more easily with prefix them as `minishift-`. eg. `/etc/resolver/192.168.64.13.xip.io` would ALWAYS imply `domain 192.168.64.13.xip.io`, but rather for convenience not name them this way.

**Comment by gbraad** [comment]
@mohammedzee1000 proposed a PR for CCCP: https://github.com/kbsingh/cccp-index/pull/45

**Comment by gbraad** [comment]
@mohammedzee1000 added https://github.com/CentOS/container-index/pull/299

**Comment by mohammedzee1000** [comment]
@gbreed. Merged. You should have the container shortly.

**Comment by gbraad** [comment]
@mohammedzee1000 not seeing my container at registry.centos.org

**Comment by gbraad** [comment]
On Fedora I need to force `systemctl restart NetworkManager` before I see the DNS change in `/etc/resolv.conf`.

**Comment by gbraad** [comment]
Proposed an intitial version for the Docker implementation: https://github.com/minishift/minishift/pull/2076

```
$ minishift image export registry.centos.org/minishift/dnsmasq:latest
$ minishift image import registry.centos.org/minishift/dnsmasq:latest
```

**Comment by gbraad** [comment]
Running in the container as systemd service: https://github.com/gbraad/dockerfile-dnsmasq/commit/bac969250365db1239f0d45e2cd2bdc30f1bf559

**Comment by gbraad** [comment]
Minikube runs `systemd-resolved`. This needs to be disabled:

```
$ sudo systemctl disable systemd-resolved.service
$ sudo systemctl stop systemd-resolved
```

**Comment by gbraad** [comment]
Added dnsmasq to the RHEL and CentOS iso as a fallback, since for RHEL we can not offer a container based deployment in a reasonable timeframe.

**Comment by gbraad** [comment]
Needs https://github.com/minishift/minishift-centos-iso/pull/232 to operate on RHEL and CentOS

**Comment by gbraad** [comment]
Enabled DNS as experimental feature and added some config options
